### PR TITLE
Improve deleting a single asset in WebUI

### PR DESCRIPTION
**Pull request by Amrysliu**
See https://progress.opensuse.org/issues/55634

### CODE REVIEW

File: assets/javascripts/admin_assets.js [hunk 1]
```diff
@@ -30,6 +30,9 @@ function setupAdminAssets() {
                     $('#assets-status').text('failed to load');
                 }
             },
+	    createdRow: function (row, data) {
+	        $(row).attr("id", "asset_" + data.id);
```
Comment (Martchus) [inline assets/javascripts/admin_assets.js]: I would write it as `row.id = 'asset_' + data.id`. We don't need to use jQuery for every tiny thing.

File: assets/javascripts/admin_assets.js [hunk 2]
```diff
@@ -117,11 +120,15 @@ function deleteAsset(assetId) {
     $.ajax({
         url: '/api/v1/assets/' + assetId,
         method: 'DELETE',
+	dataType: 'json',
         success: function() {
-            $('#asset_' + assetId).remove();
+	    var table = $("#assets").DataTable();
+	    table.row($("#asset_" + assetId)).remove().draw();
```
Comment (Martchus) [inline assets/javascripts/admin_assets.js]: Maybe we shouldn't even remove the row here. It would be only a temporary removal which does not persist reloading the page. That's why you added the flash message and trying to fake the view update contradicts that message.

File: assets/javascripts/admin_assets.js [hunk 3]
```diff
@@ -117,11 +120,15 @@ function deleteAsset(assetId) {
...
+	    addFlash('info', 'The asset was deleted successfully. Because of the asset cache, the data are shown in the table may be not effective. To make sure the data are effective, please click "Trigger asset cleanup" button.');
```
Comment (Martchus) [inline assets/javascripts/admin_assets.js]: I would rephrase it as:

> The asset was deleted successfully. The asset table's contents are cached. Hence the removal is not immediately visible. To update the view use the "Trigger asset cleanup" button. Note that this is an expensive operation which might take a while.

File: assets/javascripts/admin_assets.js [hunk 4]
```diff
@@ -117,11 +120,15 @@ function deleteAsset(assetId) {
...
         },
         error: function(xhr, ajaxOptions, thrownError) {
-            window.alert('The asset couldn\'t be deleted: ' + thrownError);
+	    var error_message = xhr.responseJSON.error;
+	    addFlash('danger', 'The asset couldn\'t be deleted: ' + error_message);
```
Comment (Martchus) [inline assets/javascripts/admin_assets.js]: Does this message show up when trying to delete an asset which has already been removed before (but was still visible in the table)? Considering the code below (`$self->render(json => {error => 'Asset not found.'}, status => 404);`) it looks like this would be the case. Maybe we should note here that the asset might have been removed already.

File: t/api/02-assets.t [hunk 5]
```diff
@@ -137,7 +137,6 @@ la;
 
 # test delete operation
 $t->delete_ok('/api/v1/assets/' . $listing->[0]->{id})->status_is(200);
-is($t->tx->res->json->{count}, 1, "one asset deleted");
```
Comment (Martchus) [inline t/api/02-assets.t]: This is actually a breaking API change. Do we care? Maybe you can let it still return the number of deleted assets for the sake of compatibility (and possibly consistency with other API routes)?

But I agree of course that the `count` makes no sense if we only support removing one asset at a time and return an error if that didn't work.

Comment (kalikiana) [inline t/api/02-assets.t]: Why are you removing those checks?

Comment (Martchus) [reply t/api/02-assets.t]: Because she changed the API and that actually makes sense if there weren't the concern about backward compatibility. (See my review comment.)

File: assets/javascripts/admin_assets.js (hunk 3 above)
Comment (okurz) [reply assets/javascripts/admin_assets.js]: I would say we do not even know if the actual deletion was successful. We only know that it was triggered, isn't it?

File: lib/OpenQA/WebAPI/Controller/API/V1/Asset.pm [hunk 6]
```diff
@@ -176,15 +175,21 @@ sub delete {
...
     }
     else {
-        return;
+        return undef;
+    }
+
+    my $asset = $self->schema->resultset("Assets")->find(\%cond);
+
+    return $self->render(json => {error => 'Asset not found.'}, status => 404) unless $asset;
+
```
Comment (okurz) [inline lib/OpenQA/WebAPI/Controller/API/V1/Asset.pm]: would you mind deleting these blank lines?

File: assets/javascripts/admin_assets.js (hunk 3 above)
Comment (Martchus) [reply assets/javascripts/admin_assets.js]: We know that is was successful.

File: t/api/02-assets.t (hunk 5 above)
Comment (Martchus) [reply t/api/02-assets.t]: I've just read the code again and noticed that we actually support deleting multiple assets at the same time. So the count should stay in my opinion.

File: assets/javascripts/admin_assets.js (hunk 1 above)
Comment (Amrysliu) [reply assets/javascripts/admin_assets.js]: As your comment below, if we do not remove the row which the deleted asset belongs to, the code is not needed.

File: lib/OpenQA/WebAPI/Controller/API/V1/Asset.pm (hunk 6 above)
Comment (Amrysliu) [reply lib/OpenQA/WebAPI/Controller/API/V1/Asset.pm]: ok, I deleted these blank lines.

File: assets/javascripts/admin_assets.js (hunk 4 above)
Comment (Amrysliu) [reply assets/javascripts/admin_assets.js]: in my opinion, the error_message includes 3 types of error:
1. The asset id is invalid. (it is not possible when deleting asset in WebUI, so I add a test in API test script - t/api/02-assets.t)
2. The asset does not exist. (The asset might have been removed. So I modify the error message in API - Asset.pm  as your suggestion.)
3. The error happens when deleting an asset from the database. (Seems like it almost never happens. I see other code about operating the database, there is always dealing with error, so I add it when deleting an asset).

File: t/api/02-assets.t (hunk 5 above)
Comment (Amrysliu) [reply t/api/02-assets.t]: I remained the count. But I did not find out where did the multiple deletion and how.  The parameter is id or type and name, how to pass more than one id? using rest api such as 'api/v1/asset/7,8'? If the parameters include more than one id,  what is the form of the search condition?

Comment (Martchus) [reply t/api/02-assets.t]: > But I did not find out where did the multiple deletion and how.

I suppose you're right after all. There's a unique constraint for the type and name (`__PACKAGE__->add_unique_constraint([qw(type name)]);`) so there should never be more than one row. But for the sake of compatibility and consistency with other routes it is good to keep the count anyways.

File: assets/javascripts/admin_assets.js [hunk 7]
```diff
@@ -117,11 +117,13 @@ function deleteAsset(assetId) {
     $.ajax({
         url: '/api/v1/assets/' + assetId,
         method: 'DELETE',
+        dataType: 'json',
         success: function() {
-            $('#asset_' + assetId).remove();
+          addFlash('info', 'The asset was deleted successfully. The asset table\'s contents are cached. Hence the removal is not immediately visible. To updatethe view use the "Trigger asset cleanup" button. Note that this is an expensive operation which might take a while.');
```
Comment (Martchus) [inline assets/javascripts/admin_assets.js]: A space is missing between "update" and "the".

File: lib/OpenQA/WebAPI/Controller/API/V1/Asset.pm [hunk 8]
```diff
@@ -176,15 +175,19 @@ sub delete {
...
     else {
-        return;
+        return undef;
     }
 
-    my $asset = $self->app->schema->resultset("Assets")->search(\%cond, \%attrs);
-    return unless $asset;
-    my $rs = $asset->delete_all;
+    my $asset = $self->schema->resultset("Assets")->search(\%cond);
+    return $self->render(json => {error => 'Asset not found. The asset might have been removed.'}, status => 404)
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/API/V1/Asset.pm]: Maybe "The asset might have *already* been removed and only the cached view has not bee updated yet.".

Comment (Amrysliu) [reply lib/OpenQA/WebAPI/Controller/API/V1/Asset.pm]: fixed it

File: assets/javascripts/admin_assets.js (hunk 7 above)
Comment (Amrysliu) [reply assets/javascripts/admin_assets.js]: fixed
